File: app/services/payment_callback.py
from decimal import Decimal

from sqlalchemy import select, func
from sqlalchemy.orm import Session

from app.models import Order, Payment, PaymentEvent, OutboxMessage
from app.errors import TransactionMismatchError, PaymentValidationError
from app.schemas import StripeWebhookRequest

def complete_toss_payment(
    session: Session,
    payment_id: str,
    transaction_id: str,
    amount: str,
    currency: str,
) -> Payment:
    amount = Decimal(amount)
    
    with session.begin():
        # Payment 잠금
        payment = session.scalar(
            select(Payment)
            .where(Payment.public_id == payment_id)
            .with_for_update()
        )

        if payment is None:
            raise PaymentValidationError(
                code="PAYMENT_NOT_FOUND",
                message="Payment not found.",
            )

        # Order 잠금
        order = session.scalar(
            select(Order)
            .where(Order.id == payment.order_id)
            .with_for_update()
        )

        if order is None:
            raise PaymentValidationError(
                code="ORDER_NOT_FOUND",
                message="Order not found.",
            )

        # 결제 제공자 검증
        if payment.provider != "TOSS":
            raise PaymentValidationError(
                code="PROVIDER_MISMATCH",
                message="Payment provider does not match.",
            )

        # 이미 처리된 동일 콜백인지 확인
        if payment.status == "COMPLETED":
            if payment.external_transaction_id == transaction_id:
                return payment

            raise TransactionMismatchError()

        # 처리 가능한 결제 상태인지 확인
        if payment.status != "PENDING":
            raise PaymentValidationError(
                code="INVALID_PAYMENT_STATUS",
                message="Payment is not in a processable state.",
            )

        # 주문 상태 검증
        if order.status != "PAYMENT_PENDING":
            raise PaymentValidationError(
                code="INVALID_ORDER_STATUS",
                message="Order is not in a payable state.",
            )

        # 금액 검증
        if payment.amount != amount:
            raise PaymentValidationError(
                code="AMOUNT_MISMATCH",
                message="Payment amount does not match.",
            )

        # 통화 검증
        if payment.currency != currency:
            raise PaymentValidationError(
                code="CURRENCY_MISMATCH",
                message="Payment currency does not match.",
            )

        # 결제 완료
        payment.status = "COMPLETED"
        payment.external_transaction_id = transaction_id
        payment.completed_at = func.now()

        # 주문 완료
        order.status = "PAID"

        # Payment 이벤트 기록
        record_payment_event(
            session,
            payment,
            transaction_id,
            amount,
            currency,
        )

        # Outbox 메시지 기록
        record_outbox_event(
            session,
            payment,
            transaction_id,
            amount,
            currency,
        )

        return payment

def complete_stripe_payment(
    session: Session,
    payload: StripeWebhookRequest,
) -> Payment:
    event_id = payload.id
    event_type = payload.type
    data = payload.data.object

    transaction_id = data["id"]
    payment_id = data["client_reference_id"]
    amount = Decimal(data["amount_total"]) / Decimal("100")
    currency = data["currency"].upper()
    payment_status = data["payment_status"]

    with session.begin():
        payment = session.scalar(
            select(Payment)
            .where(Payment.public_id == payment_id)
            .with_for_update()
        )

        if payment is None:
            raise PaymentValidationError(
                code="PAYMENT_NOT_FOUND",
                message="Payment not found.",
            )

        order = session.scalar(
            select(Order)
            .where(Order.id == payment.order_id)
            .with_for_update()
        )

        if order is None:
            raise PaymentValidationError(
                code="ORDER_NOT_FOUND",
                message="Order not found.",
            )

        if payment.provider != "STRIPE":
            raise PaymentValidationError(
                code="PROVIDER_MISMATCH",
                message="Payment provider does not match.",
            )

        # 이미 처리한 Stripe 이벤트인지 확인
        existing_event = session.scalar(
            select(PaymentEvent)
            .where(PaymentEvent.event_id == event_id)
        )

        if existing_event is not None:
            if existing_event.payment_id != payment.id:
                raise PaymentValidationError(
                    code="EVENT_PAYMENT_MISMATCH",
                    message="Stripe event is associated with a different payment.",
                )

            return payment

        if event_type != "checkout.session.completed":
            raise PaymentValidationError(
                code="UNSUPPORTED_EVENT",
                message="Unsupported Stripe event type.",
            )

        if payment_status != "paid":
            raise PaymentValidationError(
                code="PAYMENT_NOT_PAID",
                message="Stripe payment is not paid.",
            )

        # 이미 완료된 결제인데 새로운 이벤트가 들어온 경우
        if payment.status == "COMPLETED":
            if payment.external_transaction_id == transaction_id:
                return payment

            raise TransactionMismatchError()

        if payment.status != "PENDING":
            raise PaymentValidationError(
                code="INVALID_PAYMENT_STATUS",
                message="Payment is not in a processable state.",
            )

        if order.status != "PAYMENT_PENDING":
            raise PaymentValidationError(
                code="INVALID_ORDER_STATUS",
                message="Order is not in a payable state.",
            )

        if payment.amount != amount:
            raise PaymentValidationError(
                code="AMOUNT_MISMATCH",
                message="Payment amount does not match.",
            )

        if payment.currency != currency:
            raise PaymentValidationError(
                code="CURRENCY_MISMATCH",
                message="Payment currency does not match.",
            )

        payment.status = "COMPLETED"
        payment.external_transaction_id = transaction_id
        payment.completed_at = func.now()

        order.status = "PAID"

        record_payment_event(
            session,
            payment,
            transaction_id,
            amount,
            currency,
            event_id,
        )

        record_outbox_event(
            session,
            payment,
            transaction_id,
            amount,
            currency,
        )

        return payment

def complete_alipay_payment(
    session: Session,
    order_id: str,
    trade_no: str,
    total_amount: str,
) -> Payment:
    amount = Decimal(total_amount)

    with session.begin():
        # Payment 잠금
        payment = session.scalar(
            select(Payment)
            .where(Payment.public_id == order_id)
            .with_for_update()
        )

        if payment is None:
            raise PaymentValidationError(
                code="PAYMENT_NOT_FOUND",
                message="Payment not found.",
            )

        # Order 잠금
        order = session.scalar(
            select(Order)
            .where(Order.id == payment.order_id)
            .with_for_update()
        )

        if order is None:
            raise PaymentValidationError(
                code="ORDER_NOT_FOUND",
                message="Order not found.",
            )

        # 결제 제공자 검증
        if payment.provider != "ALIPAY":
            raise PaymentValidationError(
                code="PROVIDER_MISMATCH",
                message="Payment provider does not match.",
            )

        # 이미 처리된 동일 notify인지 확인
        if payment.status == "COMPLETED":
            if payment.external_transaction_id == trade_no:
                return payment

            raise TransactionMismatchError()

        # 처리 가능한 결제 상태인지 확인
        if payment.status != "PENDING":
            raise PaymentValidationError(
                code="INVALID_PAYMENT_STATUS",
                message="Payment is not in a processable state.",
            )

        # 주문 상태 검증
        if order.status != "PAYMENT_PENDING":
            raise PaymentValidationError(
                code="INVALID_ORDER_STATUS",
                message="Order is not in a payable state.",
            )

        # 금액 검증
        if payment.amount != amount:
            raise PaymentValidationError(
                code="AMOUNT_MISMATCH",
                message="Payment amount does not match.",
            )

        # Alipay 결제 완료
        payment.status = "COMPLETED"
        payment.external_transaction_id = trade_no
        payment.completed_at = func.now()

        # 주문 완료
        order.status = "PAID"

        # Payment 이벤트 기록
        record_payment_event(
            session,
            payment,
            trade_no,
            amount,
            payment.currency,
        )

        # Outbox 메시지 기록
        record_outbox_event(
            session,
            payment,
            trade_no,
            amount,
            payment.currency,
        )

        return payment

def record_payment_event(
    session: Session,
    payment: Payment,
    transaction_id: str,
    amount: Decimal,
    currency: str,
    event_id: str | None = None,
) -> None:
    event = PaymentEvent(
        payment_id=payment.id,
        event_id=event_id,
        event_type="PAYMENT_COMPLETED",
        payload={
            "provider": payment.provider,
            "transaction_id": transaction_id,
            "amount": str(amount),
            "currency": currency,
        },
    )

    session.add(event)

def record_outbox_event(
    session: Session,
    payment: Payment,
    transaction_id: str,
    amount: Decimal,
    currency: str,
) -> None:
    event = OutboxMessage(
        deduplication_key=f"payment-completed:{payment.public_id}",
        event_type="PAYMENT_COMPLETED",
        aggregate_type="payment",
        aggregate_id=payment.public_id,
        payload={
            "provider": payment.provider,
            "payment_id": payment.public_id,
            "transaction_id": transaction_id,
            "amount": str(amount),
            "currency": currency,
        },
        status="PENDING",
    )

    session.add(event)

File: internal/service/payment_callback.go
package service

import (
	"context"
	stderrors "errors"

	"encoding/json"

	"github.com/Hani-SCV/payment-callback-assignment/internal/database"
	"github.com/Hani-SCV/payment-callback-assignment/internal/errors"
	"github.com/Hani-SCV/payment-callback-assignment/internal/model"
	"github.com/Hani-SCV/payment-callback-assignment/internal/repository"
	"github.com/Hani-SCV/payment-callback-assignment/internal/request"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

type PaymentCallbackService struct {
	transactionManager      *database.TransactionManager
	orderRepository         *repository.OrderRepository
	paymentRepository       *repository.PaymentRepository
	paymentEventRepository  *repository.PaymentEventRepository
	outboxMessageRepository *repository.OutboxMessageRepository
}

type Repositories struct {
	Order        *repository.OrderRepository
	Payment      *repository.PaymentRepository
	PaymentEvent *repository.PaymentEventRepository
	Outbox       *repository.OutboxMessageRepository
}

type PaymentEventPayload struct {
	Provider      string          `json:"provider"`
	TransactionID string          `json:"transaction_id"`
	Amount        decimal.Decimal `json:"amount"`
	Currency      string          `json:"currency"`
}

type OutboxMessagePayload struct {
	Provider      string          `json:"provider"`
	PaymentID     string          `json:"public_id"`
	TransactionID string          `json:"transaction_id"`
	Amount        decimal.Decimal `json:"amount"`
	Currency      string          `json:"currency"`
}

func NewPaymentCallbackService(
	transactionManager *database.TransactionManager,
	orderRepository *repository.OrderRepository,
	paymentRepository *repository.PaymentRepository,
	paymentEventRepository *repository.PaymentEventRepository,
	outboxMessageRepository *repository.OutboxMessageRepository,
) *PaymentCallbackService {
	return &PaymentCallbackService{
		transactionManager:      transactionManager,
		orderRepository:         orderRepository,
		paymentRepository:       paymentRepository,
		paymentEventRepository:  paymentEventRepository,
		outboxMessageRepository: outboxMessageRepository,
	}
}

func (s *PaymentCallbackService) repositoriesWithTx(tx *gorm.DB) *Repositories {
	return &Repositories{
		Order:        s.orderRepository.WithTx(tx),
		Payment:      s.paymentRepository.WithTx(tx),
		PaymentEvent: s.paymentEventRepository.WithTx(tx),
		Outbox:       s.outboxMessageRepository.WithTx(tx),
	}
}

func createPaymentEvent(
	ctx context.Context,
	repo *repository.PaymentEventRepository,
	payment *model.Payment,
	eventID string,
	transactionID string,
	amount decimal.Decimal,
	currency string,
) error {
	payload := PaymentEventPayload{
		Provider:      payment.Provider,
		TransactionID: transactionID,
		Amount:        amount,
		Currency:      currency,
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	event := &model.PaymentEvent{
		PaymentID: payment.ID,
		EventID:   &eventID,
		EventType: "PAYMENT_COMPLETED",
		Payload:   jsonData,
	}

	return repo.Create(ctx, event)
}

func createOutboxMessage(
	ctx context.Context,
	repo *repository.OutboxMessageRepository,
	payment *model.Payment,
	transactionID string,
	amount decimal.Decimal,
	currency string,
) error {
	payload := OutboxMessagePayload{
		Provider:      payment.Provider,
		PaymentID:     payment.PublicID,
		TransactionID: transactionID,
		Amount:        amount,
		Currency:      currency,
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return errors.ErrInvalidPaymentStatus
	}

	event := &model.OutboxMessage{
		DeduplicationKey: payment.PublicID,
		EventType:        "PAYMENT_COMPLETED",
		AggregateType:    "payment",
		AggregateID:      payment.PublicID,
		Payload:          jsonData,
		Status:           model.PaymentStatusPending,
	}

	return repo.Create(ctx, event)
}

func (s *PaymentCallbackService) ProcessTossReturn(
	ctx context.Context,
	req request.TossReturnRequest,
) error {
	return s.transactionManager.WithTransaction(ctx, func(tx *gorm.DB) error {
		repos := s.repositoriesWithTx(tx)

		// Payment 잠금
		payment, err := repos.Payment.FindByPublicIDForUpdate(
			ctx,
			req.OrderID,
		)
		if err != nil {
			return err
		}

		// Order 잠금
		order, err := repos.Order.FindByIDForUpdate(
			ctx,
			payment.OrderID,
		)
		if err != nil {
			return err
		}

		// 결제 제공자 검증
		if payment.Provider != "TOSS" {
			return errors.ErrInvalidProvider
		}

		// 이미 처리된 결제 검증 (중복 콜백은 정상 처리한다.)
		if payment.Status == model.PaymentStatusCompleted {
			if payment.ExternalTransactionID != nil &&
				*payment.ExternalTransactionID == req.PaymentKey {
				return nil
			}
			return errors.ErrInvalidPaymentStatus
		}

		// 처리 가능한 결제 상태인지 확인
		if payment.Status != model.PaymentStatusPending {
			return errors.ErrInvalidPaymentStatus
		}

		// 주문 상태 검증
		if order.Status != "PAYMENT_PENDING" {
			return errors.ErrInvalidOrderStatus
		}

		// 금액 검증
		if !payment.Amount.Equal(req.Amount) {
			return errors.ErrInvalidAmount
		}

		// 통화 검증
		if payment.Currency != "KRW" {
			return errors.ErrInvalidCurrency
		}

		// 결제 완료
		if err := repos.Payment.Complete(
			ctx,
			payment.ID,
			req.PaymentKey,
		); err != nil {
			return errors.ErrInvalidPaymentStatus
		}

		// 주문 완료
		if err := repos.Order.MarkAsPaid(ctx, order.ID); err != nil {
			return errors.ErrInvalidOrderStatus
		}

		// PaymentEvent 저장
		if err := createPaymentEvent(
			ctx,
			repos.PaymentEvent,
			payment,
			req.PaymentKey,
			req.PaymentKey,
			req.Amount,
			payment.Currency,
		); err != nil {
			return err
		}

		// Outbox 저장
		if err := createOutboxMessage(
			ctx,
			repos.Outbox,
			payment,
			req.PaymentKey,
			req.Amount,
			payment.Currency,
		); err != nil {
			return err
		}

		return nil
	})
}

func (s *PaymentCallbackService) ProcessStripeWebhook(
	ctx context.Context,
	req request.StripeWebhookRequest,
) error {
	return s.transactionManager.WithTransaction(ctx, func(tx *gorm.DB) error {
		repos := s.repositoriesWithTx(tx)

		// data.object에서 결제 정보 추출
		object := req.Data.Object
		transactionID := object.ID
		paymentID := object.ClientReferenceID
		amount := decimal.NewFromInt(object.AmountTotal)
		currency := object.Currency
		paymentStatus := object.PaymentStatus

		// Stripe webhook event ID로 중복 이벤트 여부 확인
		_, err := repos.PaymentEvent.FindByEventID(ctx, req.ID)
		if err == nil {
			return nil
		}

		if !stderrors.Is(err, gorm.ErrRecordNotFound) {
			return err
		}

		// Stripe event type 확인
		if req.Type != "checkout.session.completed" {
			return errors.ErrInvalidEventType
		}

		// 결제 ID로 Payment 조회 및 row lock
		payment, err := repos.Payment.FindByPublicIDForUpdate(ctx, paymentID)
		if err != nil {
			return err
		}

		// Payment에 연결된 Order 조회 및 row lock
		order, err := repos.Order.FindByIDForUpdate(ctx, payment.OrderID)
		if err != nil {
			return err
		}

		// Provider가 STRIPE인지 검증
		if payment.Provider != "STRIPE" {
			return errors.ErrInvalidProvider
		}

		// 결제 상태 검증
		// 이미 완료된 결제인데 새로운 이벤트가 들어온 경우
		if payment.Status == model.PaymentStatusCompleted {
			if payment.ExternalTransactionID != nil &&
				*payment.ExternalTransactionID == transactionID {
				return nil
			}
			return errors.ErrInvalidPaymentStatus
		}

		// PENDING 상태가 아닌지 체크
		if payment.Status != model.PaymentStatusPending {
			return errors.ErrInvalidPaymentStatus
		}

		// Order 상태 검증
		if order.Status != "PAYMENT_PENDING" {
			return errors.ErrInvalidOrderStatus
		}

		// Stripe 결제 금액과 Payment 금액 비교
		if !payment.Amount.Equal(amount) {
			return errors.ErrInvalidAmount
		}

		// Stripe 통화와 Payment 통화 비교
		if payment.Currency != currency {
			return errors.ErrInvalidCurrency
		}

		if paymentStatus != "paid" {
			return errors.ErrInvalidPaymentStatus
		}

		// 결제 완료 처리
		if err := repos.Payment.Complete(ctx, payment.ID, transactionID); err != nil {
			return errors.ErrInvalidPaymentStatus
		}

		// Order PAID 처리
		if err := repos.Order.MarkAsPaid(ctx, order.ID); err != nil {
			return errors.ErrInvalidOrderStatus
		}

		// PaymentEvent 생성
		if err := createPaymentEvent(
			ctx,
			repos.PaymentEvent,
			payment,
			req.ID,
			transactionID,
			amount,
			currency,
		); err != nil {
			return err
		}

		// OutboxMessage 생성
		if err := createOutboxMessage(ctx, repos.Outbox, payment, transactionID, amount, currency); err != nil {
			return err
		}

		return nil
	})
}
